MVP functionality.

mpatek opened the pull request.

danielbachhuber commented on gif2html5.php:
```diff
@@ -1,12 +1,20 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
```
@mpatek Can you set your editor to leave trailing lines?

mpatek replied:
Sure, I'm using Sublime Text, so I just set: `"ensure_newline_at_eof_on_save": true`. Seems to work.

danielbachhuber replied:
#6

mpatek commented:
@danielbachhuber how does this look so far?

I'm guessing we want to use the `webhook` feature of the server app, so that our posts are non-blocking?
If so, should those webhooks be handled by `admin_post_(action)`? Maybe we should add some encoded query parameter to the webhook return URL so that we know it's legit (maybe using `wp_hash` on the attachment ID)?

danielbachhuber commented on gif2html5.php:
```diff
@@ -1,12 +1,28 @@
+
+add_action( 'init', function() {
+	$gif2html5 = Gif2Html5::get_instance();
+} );
```
Like we have with the theme, can we create some `GIF2HTML5()` function to wrap the singleton, and load the function on a hook.

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,111 @@
+		if ( ! $attachment_url ) {
+			return;
+		}
+		// TODO: Make this async
```
We should have it async by default. When you trigger the request, you can provide a callback URL the middleware will webhook too. You'll want to include a secret in the callback to provide some basic form of authentication.

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	private function get_api_settings() {
+		return get_option( 'gif2html5_options_api' );
```
Can we make the settings flat, instead of some unknown array?

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,111 @@
+
+		$result = null;
+		try {
+			$result = json_decode( $response_body, true );
```
`json_decode()` throws exceptions?

danielbachhuber commented:
> I'm guessing we want to use the webhook feature of the server app, so that our posts are non-blocking?
> If so, should those webhooks be handled by admin_post_(action)? Maybe we should add some encoded query parameter to the webhook return URL so that we know it's legit (maybe using wp_hash on the attachment ID)?

Yep yep and yep.

Good ideas on how we should add test coverage for this?

mpatek commented:
> Good ideas on how we should add test coverage for this?

For testing the `add_attachment/edit_attachment` hook, is there a way to mock the `wp_remote_post` function so that we can ensure that it's being called correctly?

Can we test the webhook callback by creating an attachment in the db, and then setting a variety of valid/invalid `POST` data, and then ensuring that the attachment postmeta was properly set?

danielbachhuber commented:
> For testing the add_attachment/edit_attachment hook, is there a way to mock the wp_remote_post function so that we can ensure that it's being called correctly?

I suspect there's a filter you can hook into to mock responses.

> Can we test the webhook callback by creating an attachment in the db, and then setting a variety of valid/invalid POST data, and then ensuring that the attachment postmeta was properly set?

:+1: Good idea.

mpatek commented:
@danielbachhuber Could you take a look at this again? I've made the request asynchronous, and added a webhook callback handler. I've added tests for  both of those. I've also implemented a `wp_get_attachment_image_attributes` filter for replacing the `src` attribute with the mp4 URL. Anything else that should be covered by this PR?

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,164 @@
+		$snapshot_url = esc_url_raw( $_POST['snapshot'] );
+		if ( $mp4_url && $snapshot_url ) {
+			$this->set_mp4_url( $attachment_id, $mp4_url );
+			$this->set_snapshot_url( $attachment_id, $snapshot_url );
```
Should we accept new `mp4` and `snapshot` values if they're already set in the database?

mpatek replied:
What if the user makes an update to the .gif? Shouldn't the mp4 be updated then?

danielbachhuber replied:
> What if the user makes an update to the .gif? Shouldn't the mp4 be updated then?

I guess so.

danielbachhuber commented on tests/test-gif2html5.php:
```diff
@@ -0,0 +1,348 @@
+<?php
+
+class Gif2Html5Test extends WP_UnitTestCase {
```
Can we rename to `Test_Gif2Html5` for consistency?

danielbachhuber commented on tests/test-gif2html5.php:
```diff
@@ -0,0 +1,348 @@
+		return true;
+	}
+
+	function testGif2Html5FunctionReturnsCorrectClass() {
```
Can we have these method names follow WordPress coding standards?

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,164 @@
+	}
+
+	private function __construct() {
+		$this->plugin_version = GIF2HTML5_VERSION;
```
Is this really necessary? What's the plugin version used for?

danielbachhuber commented:
> Could you take a look at this again? I've made the request asynchronous, and added a webhook callback handler. I've added tests for both of those. I've also implemented a wp_get_attachment_image_attributes filter for replacing the src attribute with the mp4 URL. Anything else that should be covered by this PR?

Looks solid — just some very small nits and then I think we can land the PR.

How well does it actually work with the app?

mpatek commented:
@danielbachhuber I reformatted the test function names, and took out the unused plugin version constant.

It was a little tricky, but I was able to test this out locally by running the gif2html5 server app inside of the vip-quickstart virtual server, and changing the api url setting to `http://localhost:5000/convert`. It worked well. I also added some fields to the attachment publish box like this:

![image](https://cloud.githubusercontent.com/assets/1451430/7300329/978b5a36-e9a9-11e4-9f64-fa5f4b426f59.png)

Unfortunately, I don't think the `wp_get_attachment_image` filtering is going to work. There's no filter that can override the `<img` tag returned by that function. As far as I can tell, you can only filter the attributes of the tag. What we need is a `<video ...><source="MP4_URL"></video>` tag. Thoughts about how we should proceed?

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,180 @@
+		if ( $mp4 = $this->get_mp4_url( $attachment_id ) ) {
+			?>
+			<div class="misc-pub-section misc-pub-gif2html5-mp4-url">
+				<label for="gif2html5_mp4_url"><?php _e( 'Mp4 URL', 'gif2html5' ) ?>:</label>
```
This should be `esc_html_e()`, not `_e()`. The former is appropriate escaping.

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,180 @@
+			?>
+			<div class="misc-pub-section misc-pub-gif2html5-mp4-url">
+				<label for="gif2html5_mp4_url"><?php _e( 'Mp4 URL', 'gif2html5' ) ?>:</label>
+				<input type="text" class="widefat urlfield" readonly="readonly" id="gif2html5_mp4_url" value="<?php esc_attr_e( $mp4 ) ?>"/>
```
This should be `echo esc_attr()`, not `esc_attr_e()`. The latter is a translation function and we aren't translating a string in this scope.

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,180 @@
+		add_action( 'attachment_submitbox_misc_actions', array( $this, 'action_attachment_submitbox_misc_actions' ), 20 );
+	}
+
+	public function action_attachment_submitbox_misc_actions() {
```
If the appropriate URLs aren't set, should we have some button to trigger a conversion?

mpatek replied:
@danielbachhuber it's a good idea, but what about the asynchronous nature of the conversion callback. If the user clicks on the `convert` button, should that trigger an AJAX request, and then do some polling to listen for an update?

mpatek replied:
Maybe we should set a flag  (boolean post-meta field) that  indicates that we've sent the conversion request and are waiting on a response. We unset the flag once we receive the response. In the post editor, if the flag is set, then we indicate that we're waiting for a response. If the flag is not set (and the mp4 URL doesn't exist), then we can show the 'convert' button. Does this sound good?

mpatek replied:
In addition - if the mp4 URL exists, then we could show a 'regenerate' button which would allow for regeneration of the mp4. So, by default the mp4 is not updated if it already exists (as you suggested above).

danielbachhuber replied:
Yep, makes sense to me.

danielbachhuber commented on inc/class-gif2html5.php:
```diff
@@ -0,0 +1,180 @@
+			?>
+			<div class="misc-pub-section misc-pub-gif2html5-snapshot-url">
+				<label for="gif2html5_snapshot_url"><?php _e( 'Snapshot URL', 'gif2html5' ) ?>:</label>
+				<input type="text" class="widefat urlfield" readonly="readonly" id="gif2html5_snapshot_url" value="<?php esc_attr_e( $snapshot ) ?>"/>
```
Same comment as above.

danielbachhuber commented:
> I reformatted the test function names, and took out the unused plugin version constant.

Cool. Couple more small things and then it's good to go.

> Unfortunately, I don't think the wp_get_attachment_image filtering is going to work.

Let's discuss in a separate issue #8

mpatek commented:
@danielbachhuber Now the misc-pub section has 3 possible states.

1. URLs do not exist for attachment, and response is not pending. Looks like this:

![generate-mp4-button](https://cloud.githubusercontent.com/assets/1451430/7308133/c264d4cc-e9df-11e4-996d-8900b265f007.png)

In this case, the 'Generate MP4' button does the same thing that the 'Update' button would do, since we automatically send a request for mp4s if they don't exist (and conversion response is not pending).

2. Response is pending. Looks like this:

![conversion-pending](https://cloud.githubusercontent.com/assets/1451430/7308151/ed535d66-e9df-11e4-9ccf-1b25a3b416c7.png)

Clicking on the 'Stop waiting' button will bring the user back to state 1 or state 3 depending on whether older MP4 URLs exist or not.

3. URLs exist. Looks like this:

![urls-exist](https://cloud.githubusercontent.com/assets/1451430/7308185/3d671eaa-e9e0-11e4-8624-aa3c51b649ac.png)

Clicking on the 'Regenerate MP4' button will force a new request, and takes us back to 2.

Let me know how this looks. Feel free to merge if you think it's OK.
